core/src/main/java/org/apache/iceberg/io/WriterFactory.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.deletes.EqualityDeleteWriter;
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+
+/**
+ * A factory for creating data and delete writers.
+ */
+public interface WriterFactory<T> {
```

Comment:
The assumption behind this API is that we may only change the spec and partition within a single transactions and other configs such as data/delete file format, sort order are static and known at the factory creation time.

Reply:
I think you mean we will initialize the `WriterFactory` with the given infomations:  schema, spec, properties etc. And then create different writers (include both equality/positional delete writer and data writers) for different partitions among different parallelism tasks. Yes, that's the current behevior that we `FileAppenderFactory` provide now, I think it's OK.

Reply:
Yeah, it is similar to the existing `FileAppenderFactory`.

Reply:
One difference there is that `FileAppenderFactory` accepts `FileFormat` while here I am saying that we actually know the format we want to use for data and deletes at the factory construction time. That's why I don't accept `FileFormat` as an argument for methods constructing writes.

Is there a use case when a single txn will write both Avro and Parquet data files? We can use different file formats for data and delete files. However, do we plan to do that for the same type of files within a txn?

Reply:
> That's why I don't accept FileFormat as an argument for methods constructing writes.

Yes,  we don't need the `FileFormat` to construct the data/delete file writers.

> Is there a use case when a single txn will write both Avro and Parquet data files?

For the current compute-engine + iceberg integration work,  I don't think we support writing both avro and parquet data files in the same txn.  If people really want to do this,  they will need to generate their own avro and parquet files, and finally commit those files by using Iceberg Table API. I think that's the correct practice.

> We can use different file formats for data and delete files.

Yes, that's really helpful.   I remember that the JOIN process between data files and positional delete files is a MERGE-SORT JOIN  because the both data files and delete files have sorted joined keys : <file_name, row_offset>.   If we have `N` positional delete files in the JOIN progress,  then we will allocate at least `N*128MB` memory for each row group (To read the whole row) from the positional delete files (If we use parquet),  that's quite huge memory consuming, which will lead to OutOfMemory in the real production.  In this case, the idea  file format for positional delete files is `AVRO` I think.